### PR TITLE
[Merged by Bors] - feat: port Order.SuccPred.Basic

### PR DESCRIPTION
**Pull request by arienmalec**
Port of order.succ_pred.basic

### CODE REVIEW

File: Mathlib/Order/SuccPred/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,1498 @@
...
+* `IsSuccArchimedean`: `SuccOrder` where `succ` iterated to an element gives all the greater
+  ones.
+* `IsPredArchimedean`: `PredOrder` where `pred` iterated to an element gives all the smaller
+  ones.
+
+## Implementation notes
+
+Maximal elements don't have a sensible successor. Thus the naïve typeclass
+```lean
+class naive_succ_order (α : Type*) [preorder α] :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:36]:
```suggestion
class NaiveSuccOrder (α : Type*) [Preorder α] :=
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1498 @@
...
+`max_of_succ_le`).
+The stricter condition of every element having a sensible successor can be obtained through the
+combination of `SuccOrder α` and `NoMaxOrder α`.
+
+## TODO
+
+Is `galois_connection pred succ` always true? If not, we should introduce
+```lean
+class succ_pred_order (α : Type*) [preorder α] extends succ_order α, pred_order α :=
+(pred_succ_gc : galois_connection (pred : α → α) succ)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:54]:
```suggestion
Is `GaloisConnection pred succ` always true? If not, we should introduce
```lean
class SuccPredOrder (α : Type*) [Preorder α] extends SuccOrder α, PredOrder α :=
(pred_succ_gc : GaloisConnection (pred : α → α) succ)
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,1498 @@
...
+  ⟨not_isMax_of_lt, fun ha => (le_succ a).lt_of_not_le fun h => ha <| max_of_succ_le h⟩
+#align order.lt_succ_iff_not_is_max Order.lt_succ_iff_not_isMax
+
+alias lt_succ_iff_not_isMax ↔ _ lt_succ_of_not_is_max
+
+theorem wcovby_succ (a : α) : a ⩿ succ a :=
+  ⟨le_succ a, fun _ hb => (succ_le_of_lt hb).not_lt⟩
+#align order.wcovby_succ Order.wcovby_succ
+
+theorem covby_succ_of_not_is_max (h : ¬IsMax a) : a ⋖ succ a :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:242]:
```suggestion
theorem covby_succ_of_not_isMax (h : ¬IsMax a) : a ⋖ succ a :=
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,1498 @@
...
+
+theorem succ_mono : Monotone (succ : α → α) := fun _ _ => succ_le_succ
+#align order.succ_mono Order.succ_mono
+
+theorem le_succ_iterate (k : ℕ) (x : α) : x ≤ (succ^[k]) x := by
+  conv_lhs => rw [(by simp only [Function.iterate_id, id.def] : x = (id^[k]) x)]
+  exact Monotone.le_iterate_of_le succ_mono le_succ k x
+#align order.le_succ_iterate Order.le_succ_iterate
+
+theorem is_max_iterate_succ_of_eq_of_lt {n m : ℕ} (h_eq : (succ^[n]) a = (succ^[m]) a)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:280]:
```suggestion
theorem isMax_iterate_succ_of_eq_of_lt {n m : ℕ} (h_eq : (succ^[n]) a = (succ^[m]) a)
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,1498 @@
...
+
+theorem succ_lt_succ_iff : succ a < succ b ↔ a < b := by simp
+#align order.succ_lt_succ_iff Order.succ_lt_succ_iff
+
+alias succ_le_succ_iff ↔ le_of_succ_le_succ _
+
+alias succ_lt_succ_iff ↔ lt_of_succ_lt_succ succ_lt_succ
+
+theorem succ_strict_mono : StrictMono (succ : α → α) := fun _ _ => succ_lt_succ
+#align order.succ_strict_mono Order.succ_strict_mono
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:350]:
```suggestion
theorem succ_strictMono : StrictMono (succ : α → α) := fun _ _ => succ_lt_succ
#align order.succ_strict_mono Order.succ_strictMono
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,1498 @@
...
+theorem le_pred_of_lt {a b : α} : a < b → a ≤ pred b :=
+  PredOrder.le_pred_of_lt
+#align order.le_pred_of_lt Order.le_pred_of_lt
+
+theorem le_of_pred_lt {a b : α} : pred a < b → a ≤ b :=
+  PredOrder.le_of_pred_lt
+#align order.le_of_pred_lt Order.le_of_pred_lt
+
+@[simp]
+theorem le_pred_iff_is_min : a ≤ pred a ↔ IsMin a :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:606]: Please review the entire file for `is_max`/`is_min`

File: Mathlib/Order/SuccPred/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,1498 @@
...
+
+theorem pred_lt_pred_iff : pred a < pred b ↔ a < b := by simp
+#align order.pred_lt_pred_iff Order.pred_lt_pred_iff
+
+alias pred_le_pred_iff ↔ le_of_pred_le_pred _
+
+alias pred_lt_pred_iff ↔ lt_of_pred_lt_pred pred_lt_pred
+
+theorem pred_strict_mono : StrictMono (pred : α → α) := fun _ _ => pred_lt_pred
+#align order.pred_strict_mono Order.pred_strict_mono
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:710]:
```suggestion
theorem pred_strictMono : StrictMono (pred : α → α) := fun _ _ => pred_lt_pred
#align order.pred_strict_mono Order.pred_strictMono
```

File: Mathlib/Order/SuccPred/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,1498 @@
...
+
+end LinearOrder
+
+section IsWellOrder
+
+variable [LinearOrder α]
+
+instance (priority := 100) IsWellOrder.toIsPredArchimedean [h : IsWellOrder α (· < ·)]
+    [PredOrder α] : IsPredArchimedean α :=
+  ⟨@fun a b => by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/SuccPred/Basic.lean:1457]:
```suggestion
  ⟨fun {a b} => by
```

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]

File: Mathlib/Order/SuccPred/Basic.lean (hunk 6 above)
Comment (arienmalec) [reply Mathlib/Order/SuccPred/Basic.lean]: done!

**Comment by rwbarton** [comment]
bors merge

**Pull request closed without merging**